### PR TITLE
Only pad initializer types for parameters without annotations and contextual signatures

### PR DESCRIPTION
**Pull request by Andarist**
fixes https://github.com/microsoft/TypeScript/issues/59936

### CODE REVIEW

**Comment by jakebailey** [comment]
@typescript-bot test it

**Comment by jakebailey** [comment]
@typescript-bot test it

**Comment by jakebailey** [comment]
@typescript-bot pack this

**Comment by Andarist** [comment]
> Maybe I'm misunderstanding it, but isn't this sorta kinda a revert of https://github.com/microsoft/TypeScript/pull/59183 but only in the parameter case? Is this too targeted? Or am I misunderstanding?

It's not a revert. Previously object literal types were contextually-typed by their corresponding patterns. 

So in a situation like this:
```ts
function test({ a = 1 } = {}) {}
```

The type of this `{}` ended up to be `{ a?: number }`. Since #59183 that's no longer the case - its type stays as `{}`. This fact/difference isn't particularly observable to the user but the signature computed based on it can be quite different. The signature is meant to be `(_: { a?: number }) => void`. 

Since the type of the initializer expressions changed in #59183, Anders introduced "object padding" for those initializer types - a way to go from the initializer type to a parameter type using the available binding pattern information. Such mechanism was already used for tuple types. So this was a way to unify the algorithm for both object and array patterns in parameter positions.

That new object padding introduced regressions but only because it didn't cover some cases (see https://github.com/microsoft/TypeScript/pull/59924 ) and because it's too eager in some other cases (like this one here). 

TLDR: it's not a revert, it's just an extension of those changes
